snuba/snapshots/__init__.py

```diff
@@ -0,0 +1,49 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+
+from contextlib import contextmanager
+from typing import NewType, Generator, IO, Optional, Sequence
+from dataclasses import dataclass
+
+SnapshotId = NewType("SnapshotId", str)
+
+
+@dataclass(frozen=True)
+class TableConfig:
+    """
+    Represents the snapshot configuration for a table.
+    """
+    table: str
+    columns: Optional[Sequence[str]]
+
+
+@dataclass(frozen=True)
+class SnapshotDescriptor:
+    """
+    Provides the metadata for the loaded snapshot.
+    """
+    id: SnapshotId
+    tables: Sequence[TableConfig]
+
+
+class BulkLoadSource(ABC):
+    """
+    Represent a source we can bulk load Snuba datasets from.
+    The bulk data has to be organized in tables represented through files
+    and must be able to provide a descriptor, but there is no constraint on where
+    this actually comes from.
+    """
+
+    @classmethod
+    def load(cls, path: str, product: str) -> BulkLoadSource:
```

Comment:
`path` seems a bit like unnecessarily coupling, but we can come back to that later if it's an issue. It might make sense to swap this argument order just in case, though — I think `product` is less likely to change.

Reply:
true, I should be able to put the path in a subclass